File: functions/alsa-lib/variants/go/0_alaw_to_s16_7.go
package main

func alawToS16(aVal byte) int32 {
	aVal ^= 0x55
	t := int32(aVal & 0x7f)
	if t < 16 {
		t = t*16 + 8
	} else {
		seg := (t / 16) & 0x07
		t = 0x108 + (t%16)*16
		t = t << (seg - 1)
	}
	if aVal&0x80 != 0 {
		return t
	}
	return -t
}


func main(){
    alawToS16('a')
}

File: functions/alsa-lib/variants/c/0_alaw_to_s16_7.c
#include <stdint.h>

static int alaw_to_s16_v8(unsigned char a_val)
{
	a_val ^= 0x55;
	int t = (a_val & 0x0F) << 4;
	int seg = (a_val & 0x70) >> 4;
	t += seg ? (0x100 << (seg - 1)) + 0x08 : 0x08;
	return (a_val & 0x80) ? t : -t;
}


int main(){
    alaw_to_s16_v8('a');
}